File: tests/hashers/CustomPasswordHasherTest.py
import os
import sys
import pytest

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "django_backend.settings")

from src.hashers.CustomPasswordHasher import CustomPasswordHasher


@pytest.fixture
def password():
    return "my_secure_password"


@pytest.fixture
def hasher():
    return CustomPasswordHasher()


def test_generate_salt(hasher):
    salt = hasher.salt()
    assert isinstance(salt, str)
    assert len(salt) > 0


def test_encode_and_verify(password, hasher):
    salt = hasher.salt()
    encoded = hasher.encode(password, salt)
    assert isinstance(encoded, str)
    assert len(encoded) > 0
    is_valid = hasher.verify(password, encoded)
    assert is_valid


def test_invalid_encoded_formats(hasher):
    # Empty encoded
    assert hasher.verify("irrelevant", "") is False
    # No separator
    assert hasher.verify("irrelevant", "notavalidformat") is False
    # Missing hash
    assert hasher.verify("irrelevant", "$somesalt") is False
    # Missing salt
    assert hasher.verify("irrelevant", "somehash$") is False


if __name__ == "__main__":
    pytest.main([__file__])

File: tests/hashers/CustomPasswordHasherTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

require_once __DIR__ . '/../../vendor/autoload.php';
require_once __DIR__ . '/../../src/hashers/CustomPasswordHasher.php';

class CustomPasswordHasherTest extends TestCase
{
  private string $password;

  protected function setUp(): void
  {
    $this->password = 'my_secure_password';
  }

  public function testInvalidEncodedFormats(): void
  {
    $this->assertFalse(CustomPasswordHasher::verify('irrelevant', ''));
    $this->assertFalse(CustomPasswordHasher::verify('irrelevant', 'notavalidformat'));
    $this->assertFalse(CustomPasswordHasher::verify('irrelevant', '$somesalt'));
    $this->assertFalse(CustomPasswordHasher::verify('irrelevant', 'somehash$'));
  }

  public function testGenerateSalt(): void
  {
    $salt = CustomPasswordHasher::getSalt();
    $this->assertIsString($salt);
    $this->assertNotEmpty($salt);
  }

  public function testHashAndVerify(): void
  {
    $encoded = CustomPasswordHasher::hash($this->password);
    $this->assertIsString($encoded);
    $this->assertNotEmpty($encoded);
    $isPhpValid = CustomPasswordHasher::verify($this->password, $encoded);
    $this->assertTrue($isPhpValid);
  }

  public function testVerifyPythonHash(): void
  {
    $from_py = 'd2db5f1a1c8658d87a0e696ca24bd86fba0c1ed43646a55649b714b8b9ff6b0c$558b06a41620e188';
    $isValid = CustomPasswordHasher::verify($this->password, $from_py);
    $this->assertTrue($isValid);
  }
}
